# Linear Regression
다음의 항목들에 대해서 학습합니다.

* 선형회귀(Linear Regression)
* 회귀모델 평가


## **1.환경준비**

### **(1) 라이브러리 로딩**

In [1]:
# 데이터 다루는 라이브러리
import numpy as np
import pandas as pd

# 데이터 시각화 라이브러리
import matplotlib.pyplot as plt

# 머신러닝 라이브러리
from sklearn.model_selection import train_test_split  # 데이터 분할
from sklearn.metrics import *          # 모델 평가 함수들 전체(*)

### **(2) data loading**

In [2]:
path = 'https://raw.githubusercontent.com/DA4BAM/dataset/master/Carseats.csv'

data = pd.read_csv(path)  # csv 파일을 불러올때, 지정한 칼럼의 데이터만 가져오기
data.head()

,Sales,CompPrice,Income,Advertising,Population,Price,ShelveLoc,Age,Education,Urban,US
0,9.50,138,73,11,276,120,Bad,42,17,Yes,Yes
1,11.22,111,48,16,260,83,Good,65,10,Yes,Yes
2,10.06,113,35,10,269,80,Medium,59,12,Yes,Yes
3,7.40,117,100,4,466,97,Medium,55,14,Yes,Yes
4,4.15,141,64,3,340,128,Bad,38,13,Yes,No


|	변수명	|	설명	|	구분	|
|----|----|----|
|	Sales 	|	 각 지역 판매량(단위 : 1000개)	|	Target	|
|	CompPrice 	|	지역별 경쟁사 판매가격(달러)	|	feature	|
|	Income 	|	가구당 평균 소득액(1000달러)	|	feature	|
|	Advertising 	|	 각 지역, 회사의 광고 예산(1000달러)	|	feature	|
|	Population 	|	 지역 인구수(단위 : 1000명)	|	feature	|
|	Price 	|	 자사 지역별 판매가격(달러)	|	feature	|
|	ShelveLoc 	|	 진열상태(범주 : Bad, Medium, Good)	|	feature	|
|	Age 	|	 지역 인구의 평균 연령	|	feature	|
|	Education 	|	 교육수준(범주 : 10~18)	|	feature	|
|	Urban 	|	 매장이 도심에 있는지 여부(범주 : Yes, No)	|	feature	|
|	US 	|	 매장이 미국에 있는지 여부(범주 : Yes, No)	|	feature	|


* 둘러보기

In [3]:
# 상/하위 몇개 행을 살펴 봅시다.
data.head()

,Sales,CompPrice,Income,Advertising,Population,Price,ShelveLoc,Age,Education,Urban,US
0,9.50,138,73,11,276,120,Bad,42,17,Yes,Yes
1,11.22,111,48,16,260,83,Good,65,10,Yes,Yes
2,10.06,113,35,10,269,80,Medium,59,12,Yes,Yes
3,7.40,117,100,4,466,97,Medium,55,14,Yes,Yes
4,4.15,141,64,3,340,128,Bad,38,13,Yes,No


In [4]:
# 각 칼럼의 타입을 살펴 봅시다.
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Sales        400 non-null    float64
 1   CompPrice    400 non-null    int64  
 2   Income       400 non-null    int64  
 3   Advertising  400 non-null    int64  
 4   Population   400 non-null    int64  
 5   Price        400 non-null    int64  
 6   ShelveLoc    400 non-null    object 
 7   Age          400 non-null    int64  
 8   Education    400 non-null    int64  
 9   Urban        400 non-null    object 
 10  US           400 non-null    object 
dtypes: float64(1), int64(7), object(3)
memory usage: 34.5+ KB


## **2. 단순회귀**

### **(1) 데이터 전처리**

#### **1) 데이터분할1 : x, y 나누기**

In [6]:
target = 'Sales'
feature = ['Price']
x = data.loc[:, feature]
y = data.loc[:, target]

#### **2) 데이터분할2 : train : validation 나누기**

In [11]:
x_train, x_val, y_train, y_val = train_test_split(x, y, test_size = .3,
                                                  random_state = 20)

### **(2) 모델링**

* 하나의 feature로 tareget 예측
    * Price --> Sales

In [7]:
from sklearn.linear_model import LinearRegression

#### **1) 모델선언**

In [8]:
model1 = LinearRegression()

#### **2) 모델링(학습)**

In [12]:
model1.fit(x_train, y_train)

LinearRegression()

* 회귀계수 확인하기

In [13]:
print('회귀계수 : ', model1.coef_)
print('절편 : ', model1.intercept_)

회귀계수 :  [-0.0524425]
절편 :  13.644386014375812


#### **3) 검증 : 예측**

In [14]:
pred1 = model1.predict(x_val)

#### **4) 검증 : 평가**

In [15]:
# 평균오차
mean_absolute_error(y_val, pred1)

2.166652749231526

## **3. 회귀 모델 평가하기**

### **(1) 오차의 비로 평가하기 : R2 Score**

In [16]:
# r2 score로 평가하고 결과를 해석해 봅시다.
r2_score(y_val, pred1)

0.18003797156441326

### **(2) 오차의 양과 율로 평가하기**

In [17]:
# MSE
mean_squared_error( y_val   , pred1   )

6.8299260510695206

In [18]:
# RMSE
root_mean_squared_error( y_val   , pred1)

2.613412721150167

In [19]:
# MAE
mean_absolute_error(y_val   , pred1)

2.166652749231526

In [20]:
# MAPE
mean_absolute_percentage_error(y_val   , pred1)

147962928206494.2

## **4. 다중회귀**

### **(1) 데이터 전처리**

#### **1) 데이터분할1 : x, y 나누기**

In [21]:
target = 'Sales'
x = data.drop(target, axis=1)
y = data.loc[:, target]

#### **2) 가변수화**
모든 범주형 변수에 대해서 가변수화를 수행합니다.

In [22]:
cat_cols = ['ShelveLoc', 'US','Urban']

In [23]:
x = pd.get_dummies(x, columns=cat_cols, drop_first=True)
x.head()

,CompPrice,Income,Advertising,Population,Price,Age,Education,ShelveLoc_Good,ShelveLoc_Medium,US_Yes,Urban_Yes
0,138,73,11,276,120,42,17,False,False,True,True
1,111,48,16,260,83,65,10,True,False,True,True
2,113,35,10,269,80,59,12,False,True,True,True
3,117,100,4,466,97,55,14,False,True,True,True
4,141,64,3,340,128,38,13,False,False,False,True


#### **3) 데이터분할2 : train : validation 나누기**

In [24]:
x_train, x_val, y_train, y_val = train_test_split(x, y, test_size = .3, random_state = 20)

### **(2) 모델링1**
Price + Age

In [25]:
# 학습에 사용할 변수를 지정합니다.
features = ['Price','Age' ]
x_train1 = x_train[features]
x_val1 = x_val[features]

In [26]:
x_train1.head()

,Price,Age
268,105,39
108,103,65
21,109,62
231,122,27
33,128,50


In [27]:
model1 = LinearRegression()
model1.fit(x_train1, y_train)

LinearRegression()

In [28]:
print(list(x_train1))
print(model1.coef_, model1.intercept_)

['Price', 'Age']
[-0.05668523 -0.05057859] 16.87498339774416


In [29]:
pred1 = model1.predict(x_val1)

In [30]:
print('RMSE :', root_mean_squared_error(y_val, pred1))
print('MAE  :', mean_absolute_error(y_val, pred1))
print('r2 :', r2_score(y_val, pred1))

RMSE : 2.5176666485070442
MAE  : 2.053592574723096
r2 : 0.23901833455873955


### (3) 모델링2
전체 feature 모두 사용

In [31]:
model3 = LinearRegression()
model3.fit(x_train, y_train)

LinearRegression()

In [32]:
print(list(x_train))
print(model3.coef_, model3.intercept_)

['CompPrice', 'Income', 'Advertising', 'Population', 'Price', 'Age', 'Education', 'ShelveLoc_Good', 'ShelveLoc_Medium', 'US_Yes', 'Urban_Yes']
[ 9.14550155e-02  1.52896750e-02  1.11545812e-01  4.10916196e-04
 -9.49346042e-02 -4.32955626e-02 -2.39647930e-02  4.75921210e+00
  1.82116013e+00 -1.00667072e-01  9.23749941e-02] 5.7885182934110055


In [33]:
pred3 = model3.predict(x_val)

In [34]:
print('RMSE :', root_mean_squared_error(y_val, pred3))
print('MAE  :', mean_absolute_error(y_val, pred3))
print('r2 :', r2_score(y_val, pred3))

RMSE : 1.1271151463198543
MAE  : 0.9415085484647864
r2 : 0.8474843845917948


## 4.복습문제


### (1) Data Loading

In [35]:
data_path = 'https://raw.githubusercontent.com/DA4BAM/dataset/master/boston.csv'
data = pd.read_csv(data_path)

**변수설명**
* medv  : 1978 보스턴 주택 가격, 506개 타운의 주택 가격 중앙값 (단위 1,000 달러) <== Target
* crim	범죄율
* zn	25,000 평방피트를 초과 거주지역 비율
* indus	비소매상업지역 면적 비율
* chas	찰스강변 위치(범주 : 강변1, 아니면 0)
* nox	일산화질소 농도
* rm	주택당 방 수
* age	1940년 이전에 건축된 주택의 비율
* dis	직업센터의 거리
* rad	방사형 고속도로까지의 거리
* tax	재산세율
* ptratio	학생/교사 비율
* lstat	인구 중 하위 계층 비율


### (2) 데이터 전처리

#### 1) x, y 분할

In [36]:
target = 'medv'
x = data.drop(target, axis=1)
y = data.loc[:, target]

#### 2) 가변수화
* 이 데이터는 범주형 변수가 chas 인데, 이미 0,1로 되어 있음 --> 가변수화 불필요

#### 3) train, val 분할

In [37]:
x_train,x_val,y_train,y_val = train_test_split(x,y, test_size=0.3)

### (3) 모델링

#### 1) 모델 선언

In [38]:
model = LinearRegression()

#### 2) 학습

In [39]:
model.fit(x_train, y_train)

LinearRegression()

* 회귀 계수 열어보기

#### 3) 예측

In [41]:
# 회귀계수
print(model.coef_)
# y절편
print(model.intercept_)

[-1.32178455e-01  4.89602397e-02  1.60667513e-02  2.07824051e+00
 -1.80024352e+01  3.56628293e+00 -1.52541406e-03 -1.43488654e+00
  3.24957675e-01 -1.52117306e-02 -1.06700572e+00 -5.02757546e-01]
44.356212547132515


In [42]:
pred = model.predict(x_val)

#### 4) 평가
* 성능 지표 : rmse, mae, mape, r2
* 성능 평가 결과를 보고 의미를 설명해 봅시다.

In [43]:
#RMSE
print(root_mean_squared_error(y_val, pred))
# MAE
print(mean_absolute_error(y_val, pred))
# MAPE
print(mean_absolute_percentage_error(y_val, pred))
# R^2
print(r2_score(y_val,pred))
#

5.117192338427436
3.420683090278531
0.1585797375463475
0.6814096827678291


##평가에 대한 설명



1.   집값은 대략 위아래로 3420달러 정도의 오차를 보인다.
2.   평균 오차율(MAPE)는 약 15.8%이다
3.   모델의 성능은 대략 68%로 모델의 성능이 압도적으로 좋진 않지만 준수한 편임을 알 수 있다.

